rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+		// (validatorSet[], index, blockNumber)
+		index, _ := castParams["index"].(uint64)
+		blockNumber, _ := castParams["blockNumber"].(uint64)
+
+		// Save to the local database
+		if err := s.storage.WriteLastProcessedEvent(
+			fmt.Sprintf("%d:%d", index, blockNumber),
+			contractAddr.String(),
+		); err != nil {
+			s.logger.Error(
+				fmt.Sprintf(
+					"Unable to save last processed event for contract %s, %v",
+					contractAddr,
+					err,
+				),
+			)
+
+			return
+		}
+
+		// Realign the local SAMP
+		s.samp.Prune(index)
+	default:
+		s.logger.Error("Unknown payload type")
+
+		return
+	}
+}
+
+// GetReadyTransaction retrieves the ready SAMP transaction which has
+// enough valid signatures
+func (s *SAMUEL) GetReadyTransaction() *types.Transaction {
+	// Get the latest verified SAM
+	verifiedSAM := s.samp.Peek()
+	if verifiedSAM == nil {
+		return nil
+	}
+
+	// Find the verified SAM that has the least quorum signatures
+	verifiedSAM = s.getVerifiedSAMBucket(verifiedSAM)
+	if verifiedSAM == nil {
+		return nil
+	}
+
+	// Extract the required data
+	SAM := []rootchain.SAM(verifiedSAM)[0]
+
+	blockNumber := SAM.BlockNumber
+	childBlockNumber := SAM.ChildBlockNum
+	index := SAM.Index
+	signatures := verifiedSAM.Signatures()
+
+	// Extract the payload info
+	payloadType, payloadData := SAM.Payload.Get()
+	rawPayload, err := getEventPayload(payloadData, uint64(payloadType))
+
+	if err != nil {
+		s.logger.Error(
+			fmt.Sprintf(
+				"Unable to extract event payload for SAM %s, %v",
+				SAM.Hash.String(),
+				err,
+			),
+		)
+	}
+
+	switch payloadType {
+	case rootchain.ValidatorSetPayloadType:
+		// Get the validator set info
+		vs, _ := rawPayload.(*payload.ValidatorSetPayload)
+		setInfo := vs.GetSetInfo()
+		validatorSetMap := make([]map[string][]byte, len(setInfo))
+
+		for index, info := range setInfo {
+			validatorSetMap[index] = map[string][]byte{
+				"Address":      info.Address,
+				"BLSPublicKey": info.BLSPublicKey,
+			}
+		}
+
+		// The method should have the signature
+		// methodName(validatorSet tuple[], index uint64, blockNumber uint64, signatures [][]byte)
+		encodedArgs, err := s.eventData.methodABI.Inputs.Encode(
+			map[string]interface{}{
+				"validatorSet":         validatorSetMap,
+				"index":                index,
+				"blockNumber":          blockNumber,
+				"signatures":           signatures,
+				"signatureBlockNumber": childBlockNumber,
+			},
+		)
+
+		if err != nil {
+			s.logger.Error(
+				fmt.Sprintf(
+					"Unable to encode method arguments for SAM %s, %v",
+					SAM.Hash.String(),
+					err,
+				),
+			)
+
+			return nil
+		}
+
+		return &types.Transaction{
+			Nonce:    0,
+			From:     types.ZeroAddress,
+			To:       &s.eventData.localAddress,
+			GasPrice: big.NewInt(0),
+			Gas:      framework.DefaultGasLimit,
+			Value:    big.NewInt(0),
+			V:        big.NewInt(1), // it is necessary to encode in rlp,
+			Input:    append(s.eventData.methodABI.ID(), encodedArgs...),
+		}
```

Comment:
You might want to define a function to create the transaction from payload because this method is too long

Reply:
I'm trying to figure out how this can be reworked to be cleaner, but I cannot find a smart way to do it - because you have these constraints to pass out many arguments. Apart from defining intermediary structs for packaging them up, I don't see a way to have something smart.

Do you have any suggestions?

Reply:
I meant we should write long code in switch-case

Reply:
The switch case code is what I was talking about in the first place - you need to transfer the extracted arguments into it regardless.

Resolving this because we might tackle it in a future PR